man/megatec_usb.8
.TH MEGATEC_USB 8 "Sun Mar 4 2007" "" "Network UPS Tools (NUT)"
.SH NAME
megatec_usb \- Driver for Megatec protocol based USB UPS equipment
.SH NOTE
This man page only documents the hardware\(hyspecific features of the
megatec_usb driver. For information about the core driver, see
\fBnutupsdrv\fR(8).

\fBNote.\fR This driver is experimental at the moment. It is currently being 
worked on and may change over time. However, it is stable in execution and 
does what is needed.

.SH SUPPORTED HARDWARE

The megatec_usb driver is known to work with various UPSes from Krauler, Agiler,
and Upsonic (Ablerex interface). Others that use the Megatec protocol
may work, but haven't been tested.

This driver supports the USB versions of UPSs, if you have only serial interface then use the
straight \fBmegatec\fR(8) driver.

If you find that this driver does not work, and you are sure that your hardware speaks megatec protocol,
then you can easily extend this driver. However, you will need to do some analysis of the 
commands that are sent to the UPS. Please see the forums for discussions on this and how
to extend this driver.

.SH EXTENSIONS TO MEGATEC

This driver extends the standard \fBmegatec\fR(8) driver to work with certain UPSes that have 
a USB interface. These devices do not obey the normal rules for HID USB devices and therefore the 
\fBnewhidups\fR(8) driver will not work for them.

This driver takes the standard serial interface and passes this over USB to the UPS.
This means that most, but not all, commands available for the serial driver will 
work. For Krauler, Agiler, and Upsonic (with Ablerex interface) UPSes 
it appears to give very similar results to the serial interface. The
driver supports monitoring most of the usual UPS variables, and also
supports automatic shutdown. 

.SH EXTRA ARGUMENTS

.IP "mfr=\fIstring\fR"

Optional.  Specify the UPS manufacturer name.

.IP "model=\fIstring\fR"

Optional.  Specify the UPS model name.

.IP "serial=\fIstring\fR"

Optional.  Specify the UPS serial number.

.IP "lowbatt=\fIfloat\fR"

Optional.  Low battery level (%). Overrides the hardware default level.

.IP "ondelay=\fIint\fR"

Optional.  Delay before the UPS is turned back on (minutes).

.IP "offdelay=\fIint\fR"

Optional.  Delay before the UPS is turned off (minutes).

.IP "battvolts=\fIfloat\fR:\fIfloat\fR"

Optional.  The battery voltage interval <empty:full> (volts).

.IP "vendor=\fIregex\fR"
.IP "product=\fIregex\fR"
.IP "vendorid=\fIregex\fR"
.IP "productid=\fIregex\fR"

Select a specific UPS, in case there is more than one connected via
USB. Each option specifies an extended regular expression (see
\fBregex(7)\fR) that must match the UPS's entire vendor/product/serial
string (minus any surrounding whitespace), or the whole 4-digit
hexadecimal code for vendorid and productid. Try \fB-DD\fR for
finding out the strings to match.

Examples: 

    \fB-x vendor="Foo.Corporation.*"\fR

    \fB-x vendorid=05b8\fR (Agiler)

    \fB-x product=".*(Smart|Back)-?UPS.*"\fR

.IP "bus=\fIregex\fR"

Select a UPS on a specific USB bus or group of busses. The argument is
a regular expression that must match the bus name where the UPS is
connected (e.g. bus="002", bus="00[2-3]"). 

.IP "subdriver=\fIstring\fR"

Select a serial-over-USB subdriver to use. You have a choice between "agiler" and "krauler" subdrivers currently.

.SH UPS COMMANDS

This driver supports some extra commands (see \fBupscmd\fR(8)):

.IP beeper.toggle
Toggle the UPS beeper. (Not available on some hardware.)

.IP load.on
Turn on the load immediately.

.IP load.off
Turn off the load immediately. (Not yet supported)

.IP reset.input.minmax
Reset minimum and maximum input voltage status. (Not yet supported)

.IP reset.watchdog
Restarts the UPS watchdog timeout. The watchdog is started if it isn't running,
which means the UPS shuts down one minute after this command is executed for
the last time, restarting one minute afterwards. To cancel the watchdog, run
the "shutdown.stop" command.

.IP shutdown.return
Turn off the load and return when power is back. (Not yet supported)

.IP shutdown.stayoff
Turn off the load and remain off. (Not yet supported)

.IP shutdown.stop
Stop a shutdown in progress.

.IP test.battery.start.deep
Perform a long battery test (Not available on some hardware.)

.IP test.battery.start
Perform a (10 second) battery test.

.IP test.battery.stop
Stop a running battery test. (Not available on some hardware.)

.SH USER VARIABLES

The behavior of the driver can be controlled by some user definable
variables (see \fBupsrw\fR(8)):

.IP ups.delay.shutdown
Interval to wait after shutdown with delay command (seconds)

.IP ups.delay.start
Interval to wait before (re)starting the load (seconds)

.SH KNOWN PROBLEMS

The battery charge (%) value is only a best-effort calculation and may be
incorrect, or non-existant, for some UPS models. In any case, you can always
resort to finding the voltages at which the battery becomes fully discharged
and fully charged, and specify them manually using the "battvolts" option.
If you do this, please notify the author of this driver.

Some UPS commands aren't supported by all models. In most cases, the driver
will send a message to the system log when the user tries to execute an
unsupported command. Unfortunately, some models don't even provide a way for
the driver to check for this, so the unsupported commands will silently
fail.

Some models report a bogus value for the beeper status (will always be
"enabled" or "disabled"). So, the "beeper.toggle" command may appear to have
no effect in the status reported by the driver when, in fact, it is working
fine.

The temperature value is also known to be bogus in some models.


.SH AUTHORS
Carlos Rodrigues <carlos.efr at mail.telepac.pt>
.br
Andrey Lelikov <nut-driver at lelik.org>
.br
Alexander Gordeev <lasaine at lvk.cs.msu.su>
.br
Jon Gough <jon.gough at eclipsesystems.com.au>

.SH SEE ALSO

\fBmegatec\fR(8), \fBnutupsdrv\fR(8), \fBupsc\fR(8), \fBupscmd\fR(8), \fBupsrw\fR(8)

.SS Internet resources:
The NUT (Network UPS Tools) home page: http://www.networkupstools.org/


NEWS
If you're upgrading from an earlier version, see the UPGRADING file.

---------------------------------------------------------------------------
Release notes for NUT 2.2.2 - what's new since 2.2.1:

 - support for new devices: APC BACK-UPS XS LCD, Atlantis Land,
   Mustek Powermust Office 650, Oneac XAU models, Powerware PW5115 and
   PW9120 (USB), Nitram Elite 2005
 
 - Integrated Power Management (NUT HAL integration) has reached a
   major milestone: it is now the most advanced UPS integration into
   Power Management layer known in existing OSs. It has received many
   corrections and improvements, and allows to PowerOff the UPS at the
   end of a power cycle (which is the most important feature, not
   supported on other systems).
   The various files are now installed into the correct location.

 - the usbhid-ups driver has received attention. Most notably, the
   shutdown handling has been reworked, and support for MGE UPS SYSTEMS
   3 phases units has been added.

 - snmp-ups now supports MGE* Environment Sensor (ref 66 846).
   The ambient.temperature reporting has also been fixed for units
   other than APC.

 - the netxml-ups driver has appeared to support MGE* network HTTP/XML
   cards.
 
 - NUT now distributes by default the shared version of libupsclient
   (version 1.0.0), and use this for the provided clients (upsmon, upsc,
   upsrw, upscmd). This is part of an effort to reduce NUT's footprint,
   both on disk and in memory.
 
 - powerpanel has reach a new step toward the replacement of nitram and
   cpsups drivers. The final step is scheduled for NUT 2.4.
 
 - many changes, cleanup and fixes to the NUT core and various drivers.

 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.2.1 - what's new since 2.2.0:

 - support for new devices: all MGE Office Protection Systems units,
   Advice TopGuard 2000, Belkin F6H375-USB, Dynamix UPS1700D, Effekta RM2000MH,
   Jageson Technology Jasuny USPS, Powercom SMK-1500A and SXL-1500A,
   PowerWalker Line-Interactive VI 400/800 and 600, Powerware 9110,
   UNITEK Alpha 2600, UPSonic CXR1000, some vintage serial APC UPSs.

 - the usbhid-ups driver has been improved, and fixed in many areas, through
   a backport of the development (trunk) version.

 - the udev rules, for Linux hotplug support of the USB UPSs, has been
   updated to support kernel newer than 2.6.22.

 - the megatec and megatec_usb drivers have also been backported from the
   development (trunk) version.
   
 - the client development files have also received some care:
   the upsclient pkg-config file has been fixed, and the upsclient.h
   file allows older NUT clients to continue using the UPSCONN structure.
   
 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.2.0 - what's new since 2.0.5:

 - The new build infrastructure, using automake, is now used.
   This has major impact on the compilation and installation procedures,
   and thus on the NUT packaging.
   For more information, refer to UPGRADING and packaging/debian/ for
   an example of migration.

 - NUT now provides support for FreeDesktop Hardware Abstraction Layer
   (HAL) which brings full Plug And Play experience to USB UPS owners.
   For more information, refer to docs/nut-hal.txt.

 - support for new devices: Ablerex 625L, ActivePower 400VA, 2000VA;
   Belkin Home Office F6H350-SER, F6H500-SER, F6H650-SER; Belkin Office
   Series F6C550-AVR; Belkin Universal UPS F6C100-UNV (USB), F6C1100-UNV
   (USB), F6C1200-UNV (USB), F6H350deUNV (serial), F6H350ukUNV (serial),
   F6H650ukUNV (serial); Compaq R3000h; Cyber Power Systems PR2200;
   Dynex DX-800U; Geek Squad GS1285U; Krauler UP-M500VA; Mecer ME-2000;
   MGE UPS SYSTEMS Ellipse MAX; Online Zinto D; PowerTech SMK-800; SVEN
   Power Pro+ series, Power Smart RM 2000; Tripp-Lite SmartOnline
   SU1500RTXL2ua, smart2200RMXL2U.

 - added IPv6 support,

 - the newmge-shut driver has appeared. This one uses the same HID core
   as usbhid-ups, but communicate over a serial link. It will eventually
   replace the current mge-shut driver.

 - client commands (upsc, upsrw and upscmd): hostname is now optional,
   and defaults to "localhost"

 - many drivers have been improved and have received bug fixes:
   powerpanel, megatec, megatec_usb, safenet, tripplite_usb, gamatronic,

 - the hotplug and udev scripts, in charge of setting the right
   permissions on the USB devices, are now installed automatically
   when appropriate.

 - more generally, the NUT core and documentation, including the manpages,
   have been improved and updated.

 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.0.5 - what's new since 2.0.4:

 This release is a backport of the development version.  Many changes
 have already been backported previously.  Thus it is more a
 synchronisation release, though it includes many bugfixes and support
 for new models.

 - support for new devices: APC Smart-UPS with 6TI firmware; Belkin
   Small Enterprise F6C1500-TW-RK; Compaq R3000 XR, R5500 XR; Cyber
   Power 550SL, 725SL, 685AVR, 800AVR, 1200AVR, AE550; Eltek; Inform
   GUARD; Microsol Rhino; Opti-UPS PowerES 420E; PowerMan RealSmart,
   BackPro; Powerware PW9315 3-phase; SOLA 305; Tripp-Lite
   SMART550USB, SMART2200RMXL2U, OMNI1000LCD, OMNI900LCD, OMNI650LCD,
   1500 LCD, AVR550U; Viewsonic PowerES 420E.

 - bcmxcp: added 3-phase support

 - megatec: better hardware support, more instant commands

 - mge-hid: support more instant commands

 - newhidups: fixed APC and Tripp Lite bugs, various memory bugs,
   improved report buffering, improved Solaris support, added
   '-x explore' option for easy diagnosis of new devices


drivers/megatec_usb.c
/* megatec_usb.c - usb communication layer for Megatec protocol based UPSes
 * 
 * Copyright (C) 2006 Andrey Lelikov <nut-driver@lelik.org>
 * Copyright (C) 2007 Alexander Gordeev <lasaine@lvk.cs.msu.su>
 * Copyright (C) 2007 Jon Gough <jon.gough at eclipsesystems.com.au>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#include "main.h"
#include "megatec.h"
#include "libusb.h"
#include "serial.h"

#include <stdio.h>
#include <limits.h>
#include <string.h>
#include <stdlib.h>

#define SUB_DRIVER_VERSION	"0.10"
#define SUB_DRIVER_NAME		"Serial-over-USB transport layer"

/* driver description structure */
upsdrv_info_t	megatec_subdrv_info = {
	SUB_DRIVER_NAME,
	SUB_DRIVER_VERSION,
	"Andrey Lelikov <nut-driver@lelik.org>" \
	"Alexander Gordeev <lasaine@lvk.cs.msu.su>" \
	"Jon Gough <jon.gough@eclipsesystems.com.au>",
	DRV_STABLE,
	{ NULL }
};

/*
    This is a communication driver for "USB HID" UPS-es which use proprietary
usb-to-serial converter and speak megatec protocol. Usually these are cheap
models and usb-to-serial converter is a huge oem hack - HID tables are bogus,
device has no UPS reports, etc.
    This driver has a table of all known devices which has pointers to device-
specific communication functions (namely send a string to UPS and read a string
from it). Driver takes care of detection, opening a usb device, string
formatting etc. So in order to add support for another usb-to-serial device one
only needs to implement device-specific get/set functions and add an entry into
KnownDevices table.

*/

static usb_communication_subdriver_t *usb = &usb_subdriver;
static usb_dev_handle *udev = NULL;
static USBDevice_t usbdevice;
static USBDeviceMatcher_t *reopen_matcher = NULL;
static USBDeviceMatcher_t *regex_matcher = NULL;

typedef struct {
	char	*name;
	int	(*get_data) (char *buffer, int buffer_size);
	int	(*set_data) (const char *str);
} subdriver_t;

/* agiler-old subdriver definition */
static int get_data_agiler_old(char *buffer, int buffer_size);
static int set_data_agiler_old(const char *str);

static subdriver_t agiler_old_subdriver = {
	"agiler-old",
	get_data_agiler_old,
	set_data_agiler_old
};

/* agiler subdriver definition */
static int get_data_agiler(char *buffer, int buffer_size);
static int set_data_agiler(const char *str);

static subdriver_t agiler_subdriver = {
	"agiler",
	get_data_agiler,
	set_data_agiler
};

/* Phoenixtec Power Co subdriver definition */
static int get_data_phoenix(char *buffer, int buffer_size);
static int set_data_phoenix(const char *str);

static subdriver_t phoenix_subdriver = {
	"phoenix",
	get_data_phoenix,
	set_data_phoenix
};

/* krauler (ablerex) subdriver definition */
static int get_data_krauler(char *buffer, int buffer_size);
static int set_data_krauler(const char *str);

static subdriver_t krauler_subdriver = {
	"krauler",
	get_data_krauler,
	set_data_krauler
};

/* list of subdrivers */
static subdriver_t *subdriver_list[] = {
	&agiler_old_subdriver,
	&agiler_subdriver,
	&krauler_subdriver,
	&phoenix_subdriver,
	NULL	/* end of list */
};

/* selected subdriver */
subdriver_t *subdriver = NULL;

typedef struct {
	int vid;
	int pid;
	subdriver_t *subdriver;
} usb_ups_t;

/* list of all known devices */
static usb_ups_t KnownDevices[] = {
	{0x05b8, 0x0000, &agiler_subdriver},	/* Agiler UPS */
	{0x0001, 0x0000, &krauler_subdriver},	/* Krauler UP-M500VA */
	{0xffff, 0x0000, &krauler_subdriver},	/* Ablerex 625L USB */
	{0x0665, 0x5161, &phoenix_subdriver},	/* Belkin F6C1200-UNV */
	{0x06da, 0x0003, &phoenix_subdriver},	/* Mustek Powermust */
	{0x0f03, 0x0001, &phoenix_subdriver},	/* Unitek Alpha 1200Sx */
	{-1, -1, NULL}		/* end of list */
};

static int subdriver_match_func(USBDevice_t *d, void *privdata)
{
	usb_ups_t *p;

	if(getval("subdriver"))
		return 1;

	for (p = KnownDevices; p->vid != -1; p++) {
		if ((p->vid == d->VendorID) && (p->pid == d->ProductID)) {
			subdriver = p->subdriver;
			return 1;
		}
	}

	return 0;
}

static USBDeviceMatcher_t subdriver_matcher = {
	subdriver_match_func,
	NULL,
	NULL
};

static void usb_open_error(const char *port)
{
	fatalx(EXIT_FAILURE, 
"No supported devices found. Please check your device availability with 'lsusb'\n"
"and make sure you have an up-to-date version of NUT. If this does not help,\n"
"try running the driver with at least 'vendorid' and 'subdriver' options\n"
"specified. Please refer to the man page for details about these options\n"
"(man 8 megatec_usb).\n"
"Please report your results to the NUT user's mailing list\n"
"<nut-upsuser@lists.alioth.debian.org>.\n"
		);
}

/* FIXME: Fix "serial" variable (which conflicts with "serial" variable in megatec.c) */
void megatec_subdrv_makevartable()
{
	addvar(VAR_VALUE, "vendor", "Regular expression to match UPS Manufacturer string");
	addvar(VAR_VALUE, "product", "Regular expression to match UPS Product string");
	/* addvar(VAR_VALUE, "serial", "Regular expression to match UPS Serial number"); */
	addvar(VAR_VALUE, "vendorid", "Regular expression to match UPS Manufacturer numerical ID (4 digits hexadecimal)");
	addvar(VAR_VALUE, "productid", "Regular expression to match UPS Product numerical ID (4 digits hexadecimal)");
	addvar(VAR_VALUE, "bus", "Regular expression to match USB bus name");
	addvar(VAR_VALUE, "subdriver", "Serial-over-USB subdriver selection");
}

int ser_open(const char *port)
{
	char *regex_array[6];
	int ret;

	char *subdrv = getval("subdriver");
	char *vid = getval("vendorid");
	char *pid = getval("productid");
	char *vend = getval("vendor");
	char *prod = getval("product");

	/* pick up the subdriver name if set explicitly */
	if(subdrv)
	{
		subdriver_t **p;

		if(!vid && !pid && !vend && !prod)
		{
			upslogx(LOG_WARNING, "It's unsafe to select a subdriver but not specify device!\n"
				"Please set some of \"vendor\", \"product\", \"vendorid\", \"productid\""
				" variables.\n");
		}

		for (p = subdriver_list; *p; p++)
		{
			if (!strcasecmp(subdrv, (*p)->name))
			{
				subdriver = *p;
				break;
			}
		}

		if(!subdriver)
			fatalx(EXIT_FAILURE, "No subdrivers named \"%s\" found!", subdrv);
	}

	/* FIXME: fix "serial" variable */
        /* process the UPS selection options */
	regex_array[0] = vid;
	regex_array[1] = pid;
	regex_array[2] = vend;
	regex_array[3] = prod;
	regex_array[4] = NULL; /* getval("serial"); */
	regex_array[5] = getval("bus");

	ret = USBNewRegexMatcher(&regex_matcher, regex_array, REG_ICASE | REG_EXTENDED);
	if (ret == -1) {
		fatal_with_errno(EXIT_FAILURE, "USBNewRegexMatcher");
	} else if (ret) {
		fatalx(EXIT_FAILURE, "invalid regular expression: %s", regex_array[ret]);
	}
	/* link the matchers */
	regex_matcher->next = &subdriver_matcher;

	ret = usb->open(&udev, &usbdevice, regex_matcher, NULL);
	if (ret < 0)
		usb_open_error(port);

	/* create a new matcher for later reopening */
	ret = USBNewExactMatcher(&reopen_matcher, &usbdevice);
	if (ret) {
		fatal_with_errno(EXIT_FAILURE, "USBNewExactMatcher");
	}
	/* link the matchers */
	reopen_matcher->next = regex_matcher;

	/* NOTE: This is here until ser_flush_io() is used in megatec.c */
	ser_flush_io(0);

	return 0;
}

int ser_set_speed(int fd, const char *port, speed_t speed)
{
	return 0;
}

int ser_set_dtr(int fd, int state)
{
	return 0;
}

int ser_set_rts(int fd, int state)
{
	return 0;
}

int ser_flush_io(int fd)
{
	char flush_buf[256];
	int i;

	/* flush input buffers */
	for (i = 0; i < 10; i++) {
		if (subdriver->get_data(flush_buf, sizeof(flush_buf)) < 1)
			break;
	}

	return 0;
}

void ser_comm_fail(const char *fmt, ...)
{
}

void ser_comm_good(void)
{
}

int ser_close(int fd, const char *port)
{
	usb->close(udev);
	USBFreeExactMatcher(reopen_matcher);
	USBFreeRegexMatcher(regex_matcher);
	return 0;
}

/*!@brief Try to reconnect once.
 * @return 1 if reconnection was successful.
 */
static int reconnect_ups(void)
{
	int ret;

	upsdebugx(2, "==================================================");
	upsdebugx(2, "= device has been disconnected, try to reconnect =");
	upsdebugx(2, "==================================================");

	usb->close(udev);

	ret = usb->open(&udev, &usbdevice, reopen_matcher, NULL);
	if (ret < 1) {
		upslogx(LOG_INFO, "Reconnecting to UPS failed; will retry later...");
		udev = NULL;
		return 0;
	} else
		upslogx(LOG_NOTICE, "Successfully reconnected");

	return ret;
}

/*!@brief Report a USB comm failure, and reconnect if necessary
 * 
 * @param[in] res	Result code from libusb/libhid call
 * @param[in] msg	Error message to display
 */
void usb_comm_fail(int res, const char *msg)
{
	switch(res) {
		case -EBUSY:
			upslogx(LOG_WARNING, "%s: Device claimed by another process", msg);
			fatalx(EXIT_FAILURE, "Terminating: EBUSY");
			//upsdrv_cleanup();
			break;

		default:
			upslogx(LOG_WARNING, "%s: Device detached? (error %d: %s)", msg, res, usb_strerror());

			if(reconnect_ups()) {
				//upsdrv_initinfo();
			}
			break;
	}
}

unsigned int ser_send_pace(int fd, unsigned long d_usec, const char *fmt, ...)
{
	char buf[128];
	size_t len;
	va_list ap;
	int ret;

	if ((udev == NULL) && (! reconnect_ups()))
		return -1;

	va_start(ap, fmt);

	len = vsnprintf(buf, sizeof(buf), fmt, ap);

	va_end(ap);

	if ((len < 1) || (len >= (int) sizeof(buf))) {
		upslogx(LOG_WARNING, "ser_send_pace: vsnprintf needed more than %d bytes", (int) sizeof(buf));
		buf[sizeof(buf) - 1] = 0;
	}

	ret = subdriver->set_data(buf);
	if(ret < 0) {
		usb_comm_fail(ret, "ser_send_pace");
	}

	return ret;
}

int ser_get_line(int fd, void *buf, size_t buflen, char endchar, const char *ignset, long d_sec, long d_usec)
{
	int len;
	char *src, *dst, c;

	if ((udev == NULL) && (! reconnect_ups()))
		return -1;

	len = subdriver->get_data((char *)buf, buflen);
	if (len < 0) {
		usb_comm_fail(len, "ser_get_line");
		return len;
	}

	dst = (char *)buf;

	for (src = (char *)buf; src != ((char *)buf + len); src++) {
		c = *src;

		if (c == endchar)
			break;

		if ((c == 0) || ((ignset != NULL) && (strchr(ignset, c) != NULL)))
			continue;

		*(dst++) = c;
	}

	/* terminate string if we have space */
	if (dst != ((char *)buf + len))
		*dst = 0;

	return (dst - (char *)buf);
}

/************** minidrivers go after this point **************************/

/*
    agiler-old subdriver
*/
/*  Protocol was reverse-engineered from Windows driver
    HID tables are completely bogus
    Data is transferred out as one 8-byte packet with report ID 0
    Data comes in as 6 8-byte reports per line , padded with zeroes
    All constants are hardcoded in windows driver
*/

#define AGILER_REPORT_SIZE      8
#define AGILER_REPORT_COUNT     6
#define AGILER_TIMEOUT          5000

static int set_data_agiler_old(const char *str)
{
        unsigned char report_buf[AGILER_REPORT_SIZE];

        if (strlen(str) > AGILER_REPORT_SIZE) {
                upslogx(LOG_ERR, "set_data_agiler: output string too large");
                return -1;
        }

        memset(report_buf, 0, sizeof(report_buf));
        memcpy(report_buf, str, strlen(str));

        return usb->set_report(udev, 0, report_buf, sizeof(report_buf));
}

static int get_data_agiler_old(char *buffer, int buffer_size)
{
        int i, len;
        char buf[AGILER_REPORT_SIZE * AGILER_REPORT_COUNT + 1];

        memset(buf, 0, sizeof(buf));

        for (i = 0; i < AGILER_REPORT_COUNT; i++) {
                len = usb->get_interrupt(udev, (unsigned char *) buf + i * AGILER_REPORT_SIZE, AGILER_REPORT_SIZE, AGILER_TIMEOUT);
                if (len != AGILER_REPORT_SIZE) {
                        if (len < 0)
                                len = 0;
                        buf[i * AGILER_REPORT_SIZE + len] = 0;
                        upsdebug_hex(5, "get_data_agiler: raw dump", buf, i * AGILER_REPORT_SIZE + len);
                        break;
                }
        }

        len = strlen(buf);

        if (len > buffer_size) {
                upslogx(LOG_ERR, "get_data_agiler: input buffer too small");
                len = buffer_size;
        }

        memcpy(buffer, buf, len);
        return len;
}


/*
    Agiler serial-to-usb device.
*/

static int set_data_agiler(const char *str)
{
	return usb->set_report(udev, 0, (unsigned char *)str, strlen(str));
}

static int get_data_agiler(char *buffer, int buffer_size)
{
	return usb->get_interrupt(udev, (unsigned char *)buffer, buffer_size, 1000);
}


/*
    Phoenixtec Power Co serial-to-usb device.
*/
static char	phoenix_buffer[32];

static int set_data_phoenix(const char *str)
{
	unsigned int	count;

	memset(phoenix_buffer, '\0', sizeof(phoenix_buffer));

	snprintf(phoenix_buffer, sizeof(phoenix_buffer), "%s", str);

	if (!strcmp(phoenix_buffer, "I\r") || !strcmp(phoenix_buffer, "C\r")) {
		/* Ignore these, since they seem to lock up the connection */
		return strlen(phoenix_buffer);
	}

	for (count = 0; count < strlen(phoenix_buffer); count += 8) {

		if (usb->set_report(udev, 0, (unsigned char *)(phoenix_buffer + count), 8) < 1) {
			return -1;
		}
	}

	return strlen(phoenix_buffer);
}

static int get_data_phoenix(char *buffer, int buffer_size)
{
	int	count;

	memset(buffer, '\0', buffer_size);

	if (!strcmp(phoenix_buffer, "I\r") || !strcmp(phoenix_buffer, "C\r")) {
		/* Echo back unsupported commands */
		snprintf(buffer, buffer_size, "%s", phoenix_buffer);
		return strlen(buffer);
	}

	for (count = 8; count <= buffer_size; count += 8) {

		/* Read data in 8-byte chunks, break on a timeout */
		if (usb->get_interrupt(udev, (unsigned char *)&buffer[count-8], 8, 1000) < 0) {
			return count-8;
		}

		upsdebugx(3, "get_data_phoenix: got so far [%s]", buffer);
		upsdebug_hex(4, "get_data_phoenix", (unsigned char *)buffer, count);
	}

	upsdebugx(3, "get_data_phoenix: buffer too small");
	return -1;
}


/*
    Krauler serial-over-usb device.

    Protocol was reverse-engineered using Windows driver.
*/

#define KRAULER_MAX_ATTEMPTS_Q1		4
#define KRAULER_MAX_ATTEMPTS_F		31
#define KRAULER_MAX_ATTEMPTS_I		15

typedef struct {
	char	*str;	/* Megatec command */
	int	index;	/* Krauler string index for this command */
	char	prefix;	/* character to replace the first byte in reply */
	int	retry;	/* number of retries (1 is typically for instant commands) */
} krauler_command_t;

static krauler_command_t krauler_command_lst[] = {
	{ "T\r",  0x04, '\0', 1 },
	{ "TL\r", 0x05, '\0', 1 },
	{ "Q\r",  0x07, '\0', 1 },
	{ "C\r",  0x0b, '\0', 1 },
	{ "CT\r", 0x0b, '\0', 1 },
	{ "Q1\r", 0x03, '(',  KRAULER_MAX_ATTEMPTS_Q1 },
	{ "I\r",  0x0c, '#',  KRAULER_MAX_ATTEMPTS_I },
	{ "F\r",  0x0d, '#',  KRAULER_MAX_ATTEMPTS_F },
	{ NULL, 0, '\0', 0 }
};

/*
Still not implemented:
	0x6	T<n>	(don't know how to pass the parameter)
	0x68 and 0x69 both cause shutdown after an undefined interval
*/


/* an intermediate buffer for 1 command's output */
static char krauler_line_buf[255];
static char krauler_line_buf_len = 0;

static int set_data_krauler(const char *str)
{
	krauler_command_t *command;
	int retval = strlen(str);

	for (command = krauler_command_lst; command->str != NULL; command++) {
		int	retry;

		if (strcmp(str, command->str)) {
			continue;
		}

		upsdebugx(3, "set_data_krauler: index [%02x]", command->index);

		krauler_line_buf_len = 0;
		for (retry = 0; retry < command->retry; retry++) {
			int	res;

			res = usb->get_string(udev, command->index, krauler_line_buf, sizeof(krauler_line_buf));
			if (res < 1) {
				/* TODO: handle_error(res) */
				upsdebugx(2, "set_data_krauler: connection failure");
				return res;
			}

			/* "UPS No Ack" has a special meaning */
			if (!strcmp(krauler_line_buf, "UPS No Ack")) {
				upsdebugx(4, "set_data_krauler: retry [%s]", krauler_line_buf);
				continue;
			}

			/* Replace the first byte of what we received with the correct one */
			if(command->prefix)
				krauler_line_buf[0] = command->prefix;

			krauler_line_buf_len = res;
			return retval;
		}

		if(command->retry > 1 && retry == command->retry)
			upsdebugx(2, "set_data_krauler: too many attempts, the UPS is probably switched off!");

		return retval;
	}

	upsdebugx(4, "set_data_krauler: unknown command [%s]", str);

	/* echo the unknown command back */
	strcpy(krauler_line_buf, str);
	krauler_line_buf_len = retval;

	return retval;
}

static int get_data_krauler(char *buffer, int buffer_size)
{
	int retrieved = (buffer_size < krauler_line_buf_len) ? buffer_size : krauler_line_buf_len;
	int left = krauler_line_buf_len - retrieved;

	memcpy(buffer, krauler_line_buf, retrieved);
	memmove(krauler_line_buf, krauler_line_buf + retrieved, left);
	krauler_line_buf_len = left;

	return retrieved;
}
